### PR TITLE
Extends TreeOperator to work hierarchically

### PR DESCRIPTION
**Pull request by ecoon**
see #453 

Note this also adds a new Op, which may be useful for non-square mesh-to-mesh coupling fluxes.  It currently is only used in the new test (which starts to hint at how one could couple surface and subsurface using two operators and non-square off-diagonal blocks), operators_tree.

### CODE REVIEW

**Comment by ecoon** [comment]
merging from master, and checking regression tests.  should be ready by tomorrow unless there are review comments.

File: src/operators/Op_MeshInjection.hh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+class Op_MeshInjection : public Op {
+public:
+  Op_MeshInjection(const Schema& schema_row, const Teuchos::RCP<const AmanziMesh::Mesh>& mesh_row,
+            const Schema& schema_col, const Teuchos::RCP<const AmanziMesh::Mesh>& mesh_col,
+            const Teuchos::RCP<Epetra_Map>& injection_,
+            bool transpose_=false)
+    : Op(schema_row, schema_col, mesh_row),
+      mesh_col_(mesh_col),
+      injection(injection_),
```
Comment (lipnikov) [inline src/operators/Op_MeshInjection.hh:33]: typically, we use a_(a) instead of a(a_).

File: src/data_structures/TreeVector.cc [hunk 2]
```diff
@@ -104,6 +104,36 @@ int TreeVector::PutScalar(double scalar) {
...
 }
 
+int TreeVector::PutScalarMasterAndGhosted(double scalar) {
+  // Set all data of this node and all child nodes to scalar.
+  int ierr = 0;
+  if (data_ != Teuchos::null) {
+    ierr = data_->PutScalarMasterAndGhosted(scalar);
+    if (ierr) return ierr;
+  }
+  for (std::vector< Teuchos::RCP<TreeVector> >::iterator subvec = subvecs_.begin();
```
Comment (lipnikov) [inline src/data_structures/TreeVector.cc:114]: Optional: I use "auto" in for-loops for clarity, but it is personal.

File: src/operators/TreeOperator.cc [hunk 3]
```diff
@@ -27,78 +27,179 @@
...
+
+TreeOperator::TreeOperator(Teuchos::ParameterList& plist)
+  : TreeOperator()
+{
+  vo_ = Teuchos::rcp(new VerboseObject("TreeOperator", plist));
+  if (plist.isSublist("inverse")) set_inverse_parameters(plist.sublist("inverse"));
+}
+
+
+TreeOperator::TreeOperator(const Teuchos::RCP<const TreeVectorSpace>& row_map,
```
Comment (lipnikov) [inline src/operators/TreeOperator.cc:60]: how about row_tvs and col_tvs to be crystal clear?

File: src/operators/TreeOperator.hh [hunk 4]
```diff
@@ -108,35 +155,23 @@ class TreeOperator : public Matrix<TreeVector,TreeVectorSpace> {
...
-  
+int ApplyInverseBlockDiagonal_(const TreeVector& X, TreeVector& Y) const;
 
  private:
-  friend Impl::TreeOperator_BlockPreconditioner;
+  friend Impl::TreeOperator_BlockDiagonalPreconditioner;
+
+  Teuchos::RCP<const TreeVectorSpace> row_map_, col_map_;
+  std::size_t row_size_, col_size_;
+  Teuchos::Array<Teuchos::Array<Teuchos::RCP<TreeOperator> > > blocks_;
```
Comment (lipnikov) [inline src/operators/TreeOperator.hh:168]: consistency inside a file: either "> >" or ">>"

File: src/operators/test/AnalyticElasticityBase.hh [hunk 5]
```diff
@@ -55,6 +60,101 @@ class AnalyticElasticityBase {
...
+
+  for (int c = 0; c != ncells_owned; ++c) {
+    mesh_->cell_get_nodes(c, &nodes);
+    int nnodes = nodes.size();
+
+    for (int i = 0; i < nnodes; i++) {
+      vol_node[0][nodes[i]] += mesh_->cell_volume(c) / nnodes;
+    }
+  }
+  vol.GatherGhostedToMaster("node");
```
Comment (lipnikov) [inline src/operators/test/AnalyticElasticityBase.hh:87]: Is vol used?

File: src/operators/test/operator_stokes.cc [hunk 6]
```diff
@@ -32,14 +32,224 @@
...
 
 #include "AnalyticElasticity02.hh"
 #include "Verification.hh"
 
+
 /* *****************************************************************
 * Stokes model: exactness test.
 ***************************************************************** */
-TEST(OPERATOR_STOKES_EXACTNESS) {
+SUITE(OPERATOR_STOKES) {
```
Comment (lipnikov) [inline src/operators/test/operator_stokes.cc:44]: Make is ticket for this before merging PR. I'll also look at it.

File: src/solvers/DirectMethodAmesos.cc [hunk 7]
```diff
@@ -135,44 +135,43 @@ void DirectMethodAmesos::ComputeInverse() {
...
   Epetra_Vector* vv = const_cast<Epetra_Vector*>(&v);
   problem_->SetRHS(vv);
   problem_->SetLHS(&hv);
-  
+
   if (returned_code_ == 0) {
     int ierr = solver_->Solve();
     if (ierr) {
-      returned_code_ = -ierr;
+      returned_code_ = ierr;
```
Comment (lipnikov) [inline src/solvers/DirectMethodAmesos.cc:147]: I remember that sign was changed for a reason. But now, the return value is 0 or 1, so should be fine. correct?

File: src/operators/Op_MeshInjection.hh (hunk 1 above)
Comment (ecoon) [reply src/operators/Op_MeshInjection.hh:33]: Depends on whether the internally stored variable is public or private.  Private variables (and methods) should end in an "_" while public ones should not.

Op classes were originally structs that had no methods and no private data -- only public data for the schema, matrices, and diag.  Since then we've started to add more methods and private data, but haven't changed everything over to a more typical class, so they're halfway between classes and structs.

File: src/data_structures/TreeVector.cc (hunk 2 above)
Comment (ecoon) [reply src/data_structures/TreeVector.cc:114]: Yeah, I agree, more auto for these types of loops is a good thing.  Will change it.  A lot of the loops in TreeVector would benefit from switching to the much cleaner:

```
for (const auto& thing : container) {
  ....
}
```

notation but they were pre-C++11 code.  Will fix them all now.

File: src/operators/test/AnalyticElasticityBase.hh (hunk 5 above)
Comment (ecoon) [reply src/operators/test/AnalyticElasticityBase.hh:87]: Removed.  Hopefully you get a chance to look at that test (operators_stokes) where I tried to add a forward apply check to tests TreeOperator and couldn't get it working on master -- I am not sure why it doesn't seem to work there.  It looked like possibly the boundary conditions were the issue -- if we fill the vector in the usual way, calculate Au - rhs, then the only non-small values were on boundaries, in (if I recall correctly) the normal flux component of the vector.

Likely I just don't completely understand the system of equations being solved there, but I left it in case you wanted to take a look.

File: src/operators/TreeOperator.cc (hunk 3 above)
Comment (ecoon) [reply src/operators/TreeOperator.cc:60]: I went with map instead of tvs to be consistent across the interface.  I'd like all of our operators to have as similar an interface as possible, e.g. `get_row_map()` in both Operator and TreeOperator, as opposed to `get_row_tvs()` and `get_row_cvs()`.

So if the getters are then fixed, I think it is better to stay consistent across the variable name and the getter, e.g. if it is `get_row_map()`, then the internally stored variable should be called `row_map_` and the constructor argument should be `row_map`.

File: src/operators/test/operator_stokes.cc (hunk 6 above)
Comment (ecoon) [reply src/operators/test/operator_stokes.cc:44]: thanks

File: src/solvers/DirectMethodAmesos.cc (hunk 7 above)
Comment (ecoon) [reply src/solvers/DirectMethodAmesos.cc:147]: Yeah, we were inconsistent across the linear operators of what we returned there.  Some (our iterative methods) used the strategy of positive values indicate convergence, 0 indicated still going, and negative indicated errors.  Others used 0 = success, 1 = failure.  Still others returned 1 on success and 0 on failure (which was leftover from a previous attempt at standardizing).

Effectively I standardized all of them to return 0 on success and 1 on failure, and store their (algorithm dependent) return code, which you can get via `returned_code()` or `returned_code_string()`.  I think this is the most natural for client code who just want a "pass/fail" flag, and consistent with the rest of our interfaces which always have `ierr = 1` implies failure.  

Since this lets us allow each algorithm to have their own idea of what the code means, there was no point in negating this one, and it was potentially confusing to do so.

**Comment by ecoon** [comment]
closes #453
